Add several flow features and bugfixes

davidswinegar opened the pull request:
Add coverage checking for Flow, some bugfixes and extra error
handling. Also add two new features - the option to mix sublime
and flow autocomplete, and to omit function parameters for
autocomplete (similar to Nuclide). Add the ability to override
settings on a per-project basis under the "FlowIDE" key, so that
different projects can use different paths for Flow.

Also add an MIT license (discussed over email).

davidswinegar committed:
Add several flow features and bugfixes

Add coverage checking for Flow, some bugfixes and extra error
handling. Also add two new features - the option to mix sublime
and flow autocomplete, and to omit function parameters for
autocomplete (similar to Nuclide). Add the ability to override
settings on a per-project basis under the "FlowIDE" key, so that
different projects can use different paths for Flow.

Also add an MIT license (discussed over email).

AriaMinaei commented:
I just tried this branch. Fixes the two biggest painpoints. I'm very grateful.

I'd say the only thing remaining to bring FlowIDE up to par with Nuclide is the ability to show the flow errors in a small popup box. Because as soon as an error message gets a bit long, I'd have to run `$  flow` inside the terminal to really see what's going on. Afaik this feature is doable because Sublime added a capability that allows packages to do simple HTML rendering a while ago.

STRML commented:
:+1: Amazing. The coverage output is exactly what I wanted (barely visible but there), and the autocomplete fixes are really great. Really nice work!

davidswinegar commented:
Hey @tptee - would you be able to take a look at this?

STRML committed:
Minor flake8 cleanup

STRML committed:
Debounce on modified selection.

davidswinegar committed:
Merge pull request #1 from STRML/master

Debounced checking.

davidswinegar committed:
minor style fixes

Fix code style for accessing settings object, and remove dots
next to uncovered flow lines (just show light stipled underline).

tptee commented:
Looking now!

tptee commented on flow-ide.py:
```diff
@@ -105,16 +137,20 @@ def parse_cli_dependencies(view, **kwargs):
     try:
         output = subprocess.check_output(
-            command, stderr=subprocess.STDOUT, stdin=read
+            command, stderr=subprocess.STDOUT, stdin=read, shell=False
```
Good call!

tptee commented on flow-ide.py:
```diff
@@ -127,7 +163,32 @@ def call_flow_cli(contents, command):
             return None
+    except Exception as e:
+        os.close(read)
+        view.erase_regions('flow_error')
```
This looks copied from above. Can we move this to a utility function like `reset()` or something?

tptee commented on flow-ide.py:
```diff
@@ -149,7 +210,7 @@ def run_async(self):
             '--json',
             str(deps.row + 1), str(deps.col + 1)
-        ])
+        ], self.view)
```
I'd prefer this to be the second argument to `call_flow_cli`

tptee commented on flow-ide.py:
```diff
@@ -260,6 +328,11 @@ def on_query_completions_async(self, view, prefix, locations):
     @wait_for_load
     def on_selection_modified_async(self, view):
+        flow_settings = find_flow_settings(view.window().project_data())
+        # print('Selection modified, delaying %d' % flow_settings.get('flow_debounce_ms'))
```
Nit: commented code

tptee commented on flow-ide.py:
```diff
@@ -268,71 +341,107 @@ def on_selection_modified_async(self, view):
                 view.erase_regions('flow_error')
                 view.set_status('flow_error', 'Flow: no errors')
-                return
+            else:
```
Can we do an early `return` before this so that all the coverage code isn't indented so far?

tptee reviewed (approved):
Works wonders! I have some stylistic comments, but I think those are mainly the result of this single file getting too big. I'm going to go ahead and merge, then split out some of `flow-ide.py`.

Thanks for your contribution!! 🎉

tptee commented:
Sorry for the wait, I went down a refactoring rabbit hole 😬 `v1.2.0` is published and Package Control should pick up the update within the next 30 mins.
